nifi-nar-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/PutBoxFile.java

```diff
@@ -0,0 +1,366 @@
+                    } else {
+                        uploadedFile = parentFolder.uploadFile(rawIn, filename);
+                    }
+                }
+
+            } catch (BoxAPIResponseException e) {
+                handleUploadError(conflictResolution, filename, parentFolder, e);
+            }
+
+            if (uploadedFile != null) {
+                final Map<String, String> attributes = BoxFileUtils.createAttributeMap(uploadedFile);
+                final String url = BOX_URL + uploadedFile.getID();
+                flowFile = session.putAllAttributes(flowFile, attributes);
+                final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                session.getProvenanceReporter().send(flowFile, url, transferMillis);
+            }
+
+            session.transfer(flowFile, REL_SUCCESS);
+        } catch (BoxAPIResponseException e) {
+            getLogger().error("Exception occurred while uploading file '{}' to Box folder '{}'", filename, fullPath, e);
+            handleExpectedError(session, flowFile, e);
+        } catch (Exception e) {
+            getLogger().error("Unexpected exception occurred while uploading file '{}' to Box folder '{}'", filename, fullPath, e);
+            handleUnexpectedError(session, flowFile, e);
+        }
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        final BoxClientService boxClientService = context.getProperty(BoxClientService.BOX_CLIENT_SERVICE).asControllerService(BoxClientService.class);
+
+        boxAPIConnection = boxClientService.getBoxApiConnection();
+    }
+
+    BoxFolder getFolder(String folderId) {
+        return new BoxFolder(boxAPIConnection, folderId);
+    }
+
+    private BoxFolder.Info getOrCreateParentSubfolder(String folderName, String parentFolderId, boolean createFolder)  {
+        final int indexOfPathSeparator = folderName.indexOf("/");
+
+        if (isMultiLevelFolder(indexOfPathSeparator, folderName)) {
+            final String mainFolderName = folderName.substring(0, indexOfPathSeparator);
+            final String subFolders = folderName.substring(indexOfPathSeparator + 1);
+            final BoxFolder.Info mainFolder = getOrCreateFolder(mainFolderName, parentFolderId, createFolder);
+            return getOrCreateParentSubfolder(subFolders, mainFolder.getID(), createFolder);
+        } else {
+            return getOrCreateFolder(folderName, parentFolderId, createFolder);
+        }
+    }
+
+    private BoxFolder.Info getOrCreateFolder(String folderName, String parentFolderId, boolean createFolder) {
+        final Optional<BoxFolder.Info> existingFolder = checkFolderExistence(folderName, parentFolderId);
+
+        if (existingFolder.isPresent()) {
+            return existingFolder.get();
+        }
+
+        if (createFolder) {
+            getLogger().debug("Create folder " + folderName + " parent id: " + parentFolderId);
```

Comment:
Log statements should use placeholders:
```suggestion
            getLogger().debug("Creating Folder [{}] Parent [{}]", folderName, parentFolderId);
```